DOC-467 - add restart procedure (#1829)
hazelcast/hz-docs: docs/modules/maintain-cluster/pages/restart-cluster.adoc

{description}

== Restart options

There are two options for restarting a cluster:

* <<rolling>>
* <<whole>>

Where possible, you should perform a rolling restart. In a rolling restart, members are stopped one at a time and each member's data is automatically repartitioned across the remaining cluster members before shutting down. This minimizes the risk of service interruptions and data loss. When the replacement member joins the cluster, data is partitioned across the cluster again.

Depending on the scope of your configuration changes, you may need to update a single member, a cluster, or multiple clusters. The process is the same in all cases.

TIP: You can also use the rolling restart process for canarying: testing changes on a single member before updating the rest of the cluster.

Some configuration needs to be consistent across all members in a cluster. To update this configuration, you must restart the whole cluster simultaneously. By default, this causes all data in the cluster to be lost. To avoid data loss, you must enable xref:storage:persistence.adoc[persistence]. To avoid downtime, you must divert traffic to another Hazelcast cluster before restarting.

> Note that members are not restarted. In all cases, members are terminated, and new members are created to replace them.

Reviewer: I still think that this rather confuses than clarifies because there are different levels of member "identity" - some are preserved (address; UUID with HotRestart)

Author: Thanks for clarifying, I agree that's better.

Persistence is required to avoid data loss during a whole cluster restart. See xref:storage:configuring-persistence.adoc[] for details.

=== Embedded mode

If your cluster is operating in embedded mode, you should restart your host application instead of trying to restart the embedded Hazelcast members independently.

== Perform a restart

This section describes how to restart a cluster using a choice of common tools. Hazelcast supports a wide range of deployment options that may require other tools or allow you to automate parts of the process, but the required steps will be similar.

You should perform restarts in a low utilization or maintenance window to reduce the risk of service impacts.

[[rolling]]
=== Perform a rolling restart

You should check the cluster is healthy before and after making changes. The best way to do this is to run a xref:{page-latest-supported-mc}@management-center:clusters:healthcheck.adoc[Healthcheck] in Management Center. For other options, see xref:maintain-cluster:monitoring.adoc[].

To perform a rolling restart:

. Update the static configuration as required.

> . Choose a member and make a note of its name or IP address to help you keep track.

Reviewer: isn't this too obvious? also the procedure does not make any use of it later

. Start a new member.
+
[tabs]
====
CLI::
+
--
[source,bash]
----
hz start
----
--

Docker::
+
--
[source,bash]
----
docker start <CONTAINER_NAME>
----
--
====

. Wait until the new member joins the cluster and all partition migrations are completed. Again, you can monitor progress using the xref:{page-latest-supported-mc}@management-center:clusters:dashboard.adoc[cluster dashboard] in Management Center.

. Repeat this process until all members have been restarted with the updated configuration.

[[whole]]
=== Perform a whole cluster restart

You should check the cluster is healthy before and after making changes. The best way to do this is to run a xref:{page-latest-supported-mc}@management-center:clusters:healthcheck.adoc[Healthcheck] in Management Center. For other options, see xref:maintain-cluster:monitoring.adoc[].

To perform a whole cluster restart:

. Update the static configuration as required.

. Divert traffic to another Hazelcast cluster. How to do this depends on how your Hazelcast deployment is architected.

> +

Reviewer: IMO this looks a bit odd, single bullet point should be better:

<img width="425" height="110" alt="image" src="https://github.com/user-attachments/assets/cbce16dc-7d63-4183-9fb3-28311c4b55a3" />

Management Center provides an easy way to xref:{page-latest-supported-mc}@management-center:clusters:client-filtering.adoc[manage client connections] using filtering rules.

> . Gracefully shut down the cluster:

Reviewer: ```suggestion
. Change cluster state to `NO_MIGRATION` or `FROZEN` state as needed with Persistence
. Gracefully shut down the cluster:
```

+ example of changing the state. [Should be possible with MC and hz-cluster-admin](https://docs.hazelcast.com/hazelcast/5.5/maintain-cluster/cluster-member-states#changing-a-clusters-state).

+
[tabs]
====
CLI::
+
--
[source,bash]
----

> hz-cluster-admin -a <MEMBER_ADDRESS> -c <CLUSTER_NAME> -o shutdown

Reviewer: you can also invoke `hz-stop` on each member. `hz-cluster-admin` requires deprecated REST API

= Restart a cluster
:description: Static configuration is configuration that cannot be changed while members are running. To apply updates to static configuration, you must restart the cluster.

{description}

> There are two options for restarting a cluster:

Reviewer: this page is somewhat related to https://docs.hazelcast.com/hazelcast/5.5/maintain-cluster/shutdown - should it be linked?

Author: I don't think so - the idea is to provide a self-contained procedure for a routine operation (restarting the cluster). Unless there's any useful info in that shutdown topic that you think should be included or linked to here?

Reviewer: rolling upgrade is already described https://docs.hazelcast.com/hazelcast/5.5/maintain-cluster/rolling-upgrades#rolling-upgrade-procedure. rolling restart would be the same except version changes

whole cluster restart depends if you have persistence/hotrestart (makes sense to be graceful and there are more options during start + additional cluster states) or not (simply kill and start)

Author: Thank you, that's useful context. So we can't cover every possibility here. But I don't think it's acceptable to require a restart to change most product config and not tell users how to perform it.

The rolling upgrade procedure is poor because it doesn't tell you _how_ to do most of the steps and links out do other docs mid-flow, which users hate. I agree the method is much the same and they could arguably be combined, but I think it's useful to have dedicated docs for each - there's specific info about versions etc. in there which is useful for upgrading and a distraction for restarting.

Reviewer: > The rolling upgrade procedure is poor because it doesn't tell you how to do most of the steps and links out do other docs mid-flow, which users hate

I agree it would be better to provide some concrete examples how to perform these steps. But there are so many ways to do it depending on how you deploy the cluster (manual using hz-cli with or without some steps done using MC, systemd, ansible, helm chart, operator, embedded cluster).

Maybe we could provide docs for a basic option (hz-cli?) and other ways have their own documentation so we avoid duplication. Alternatively, maybe we should have a separate documentation belonging to particular component how to perform restart in each of the options (hz-cli, operator, maybe MC etc.) instead of putting them all together.

And actually in current state of the PR we already provide something like this. But parts of restart documentation overlap with and duplicate https://docs.hazelcast.com/hazelcast/5.5/maintain-cluster/shutdown - eg. section about how to shut down a member.

Author: Fair point. That's also duplication though, right? Do you really want 10 different topics for one procedure? We have the tabs so you can cover multiple methods in one place, which is surely more convenient for both us and users. Happy to have some level of compromise though, e.g. Operator covers it in its own docs, and if you use an orchestrator then it's reasonable to leave the detailed procedure up to the customer and just outline Hazelcast's requirements. But in principle, the docs need to explain how to use the product.

If you need to perform a shut down as a key part of a procedure, the instructions need to be in that procedure, duplication be damned. You can't send users off to another doc in the middle of a procedure, it's a real pain point. I can potentially use an `include` so it's not duplicated in the source.

* <<rolling>>
* <<whole>>

> Where possible, you should perform a rolling restart. In a rolling restart, members are stopped one at a time and each member's data is automatically repartitioned across the remaining cluster members before shutting down. This minimizes the risk of service interruptions and data loss. When the replacement member joins the cluster, data is partitioned across the cluster again.

Reviewer 2: Nit - partitioned, or repartitioned? we use both.

Author: Yeah, this is a bit of a cop out. I didn't want to say "repartitioned again" because it looks tautologous.

Depending on the scope of your configuration changes, you may need update a single member, a cluster, or multiple clusters. The process is the same in all cases.

TIP: You can also use the rolling restart process for canarying: testing changes on a single member before updating the rest of the cluster.

Some configuration needs to be consistent across all members in a cluster. To update this configuration, you must restart the whole cluster simultaneously. This causes all data in the cluster to be lost. To avoid data loss, you must enable xref:storage:persistence.adoc[persistence]. To avoid downtime, you must divert traffic to another Hazelcast cluster before restarting.

> Note that members are not restarted. In all cases, members are destroyed, and new members are created to replace them.

Reviewer: I have an impression that "destroyed" is not frequently used in context of members. Members are shutdown, terminated or leave the cluster.

> You should perform restarts in a low utilization or maintenance window to reduce the risk of service impacts.

Author: The details are up to customers but it's always worth highlighting when something is potentially service impacting.

> You should perform restarts in a low utilization or maintenance window to reduce the risk of service impacts.

Reviewer 3: This page is good, better than the rolling restart one linked above - either as part of this PR or as a quick follow-up, I'd recommend reviewing and editing rolling-restart.

== Persistence

> If xref:storage:persistence.adoc[persistence] is enabled, replacement members will adopt the UUIDs of the shut down members and attempt to load persisted data from disk. This avoids the repartitioning process, helping the cluster recover more quickly during a rolling restart.

Reviewer 2: ```suggestion
If xref:storage:persistence.adoc[persistence] is enabled, replacement members will adopt the UUIDs of the destroyed members and attempt to load persisted data from disk. This avoids the repartitioning process, helping the cluster recover more quickly during a rolling restart.
```

Based on the earlier statement:
> In all cases, members are destroyed

> If xref:storage:persistence.adoc[persistence] is enabled, replacement members will adopt the UUIDs of the destroyed members and attempt to load persisted data from disk. This avoids the repartitioning process, helping the cluster recover more quickly during a rolling restart.

Reviewer: > This avoids the repartitioning process

only if you switch the cluster to `NO_MIGRATION` or `FROZEN` state (eg. https://docs.hazelcast.com/hazelcast/5.5/kubernetes/kubernetes-persistence#configuration)

Persistence is required to avoid data loss during a whole cluster restart. See xref:storage:configuring-persistence.adoc[] for details.

[[rolling]]
== Perform a rolling restart

You should check the cluster is healthy before and after making changes. The best way to do this is to run a xref:{page-latest-supported-mc}@management-center:clusters:healthcheck.adoc[Healthcheck] in Management Center. For other options, see xref:maintain-cluster:monitoring.adoc[].

To perform a rolling restart:

. Update the static configuration as required.

. Choose a member and make a note of its name or IP address to help you keep track.

. Gracefully shut down the member.
+
[tabs]
====
CLI::
+
IMPORTANT: This command shuts down all running members on the local device.
+
--
[source,bash]
----

> bin/hz-stop

Reviewer 3: Is this correct in all cases?
I can run `hz-stop` but not `bin/hz-stop`...and I get `No Hazelcast server found to stop`

Could just be my setup or misunderstanding...

Author: Interesting. Other docs with this command specify /bin. if If you create a new member and access its terminal, you're placed at `/opt/hazelcast`, and `/bin` is a child of that directory, so it's sensible. But it looks like it's not required and hz-stop runs from anywhere, so better to omit the path. Thanks for testing.

However, if you try and run it with /bin from another directory it should just fail to find the script. Not sure how you end up with `No server found`.

----
--

Java::
+
--
[source,java]
----
HazelcastInstance.shutdown()
----
--

Management Center::
+
--
Go to *Members*, select the member's row in the table and select *Shutdown*.
--

Docker::
+
--
[source,bash]
----
docker stop <container name>
----
--
====

. Wait until all partition migrations are completed. You can monitor the progress of migrations on the xref:{page-latest-supported-mc}@management-center:clusters:dashboard.adoc[cluster dashboard] in Management Center.

. Start a new member.
+
[tabs]
====
CLI::
+
--
[source,bash]
----
bin/hz start
----
--

Java::
+
--
[source,java]
----
Hazelcast.newHazelcastInstance()
----
--

Docker::
+
--
[source,bash]
----

> docker start <container name>

Reviewer 3: I don't know if we have a set standard for placeholders but I've been favouring underscores to keep it unbroken e.g. <your_license_key> - could we agree a standard on this and keep it?

----
--
====

> . Wait until the new member joins the cluster and all partitions are completed. Again, you can monitor progress using the xref:{page-latest-supported-mc}@management-center:clusters:dashboard.adoc[cluster dashboard] in Management Center.

Reviewer 2: ```suggestion
. Wait until the new member joins the cluster and all partitions migrations are completed. Again, you can monitor progress using the xref:{page-latest-supported-mc}@management-center:clusters:dashboard.adoc[cluster dashboard] in Management Center.
```

. Repeat this process until all members have been restarted with the updated configuration.

[[whole]]
== Perform a whole cluster restart

You should check the cluster is healthy before and after making changes. The best way to do this is to run a xref:{page-latest-supported-mc}@management-center:clusters:healthcheck.adoc[Healthcheck] in Management Center. For other options, see xref:maintain-cluster:monitoring.adoc[].

To perform a whole cluster restart:

. Update the static configuration as required.

. Divert traffic to another Hazelcast cluster. How to do this depends on how your Hazelcast deployment is architected. For example, you may use a load balancer between your clients and clusters.
+
Management Center provides an easy way to xref:{page-latest-supported-mc}@management-center:clusters:client-filtering.adoc[manage client connections] using filtering rules.

. Gracefully shut down the cluster:
+
[tabs]
====
CLI::
+
--
[source,bash]
----
bin/hz-cluster-admin -a <member-address> -c <cluster-name> -o shutdown
----
--
+
You can specify the IP address of any member in the cluster.

Java::
+
--
[source,java]
----

> HazelcastInstance.getCluster().shutdown()

Reviewer: note that this does not work from client

Author: @k-jamroz Should we provide instructions for shutting down / starting from a client? (And if so, can you please provide them?)

Reviewer: Starting an instance is not possible from client. Shutting down a member/cluster is not possible using public API. Management Center has private, management API for that

----
--

Management Center::
+
--
Go to *Administration* > *Cluster State* and select *Shutdown*.
--

Docker::
+
Use `docker stop` to shut down each container as required.
+
--
[source,bash]
----
docker stop <container name>
----
--
====

. Check that all members are in `SHUT_DOWN` state using the xref:{page-latest-supported-mc}@management-center:clusters:dashboard.adoc[cluster dashboard] in Management Center.

. Recreate the cluster:
+
[tabs]
====
CLI::
+
Use `hz start` to create each new member as required. They will form a cluster automatically.
+
--
[source,bash]
----
bin/hz start
----
--

Java::
+

> Call `newHazelcastInstance` to create each new member as required. They will form a cluster automatically. The following example creates a new three member cluster.

Reviewer: the cluster will be running embedded in current process - likely not suitable for production

Author: @k-jamroz Very good point, we want this to work for production deployments. Can you provide instructions for client/server?

Reviewer: you cannot start a (standalone=non-embedded) member using client. in order to start a member you need to start a new process on the server
